src/handler/sign_tx.c

```diff
@@ -1,103 +1,423 @@
+    // buffer_read_u8(cdata, G_context.tx_info.remaining_inputs);
+    // buffer_read_u8(cdata, G_context.tx_info.outputs_len);
+
+    if (!(buffer_read_u16(cdata, &G_context.tx_info.tx_version, BE) && // read version bytes (Big Endian)
+        buffer_read_u8(cdata, &G_context.tx_info.remaining_tokens) && // read number of tokens, inputs and outputs, respectively
+        buffer_read_u8(cdata, &G_context.tx_info.remaining_inputs) &&
+        buffer_read_u8(cdata, &G_context.tx_info.outputs_len))) {
+        // if an error occurs reading
+        THROW(SW_WRONG_DATA_LENGTH);
+    }
+}
+
+void copy_sighash_all(buffer_t *cdata) {
+    // cx_hash returns the size of the hash after adding the data, we can safely ignore it
+
+    cx_hash(
+        &G_context.tx_info.sha256.header,   // hash context pointer
+        0,                                  // mode (supports: CX_LAST)
+        cdata->ptr + cdata->offset,         // Input data to add to current hash
+        cdata->size - cdata->offset,        // Length of input data
+        NULL, 0);                           // output (if flag CX_LAST was set)
+}
+
+
+bool sign_tx_with_key() {
+    // the bip32 path and bip32 path length should be on the global context when calling this method.
+
+
+    cx_ecfp_private_key_t private_key = {0};
+    cx_ecfp_public_key_t public_key = {0};
+
+    uint8_t chain_code[32];
+    uint8_t hash_160[20];
+
+    derive_private_key(&private_key, chain_code, G_context.bip32_path.path, G_context.bip32_path.length);
+    init_public_key(&private_key, &public_key);
+
+    if (G_context.tx_info.sighash_all[0] == '\0') {
+        // finish sha256 from data
+        cx_hash(&G_context.tx_info.sha256.header, CX_LAST, G_context.tx_info.sighash_all, 0, G_context.tx_info.sighash_all, 32);
+        cx_sha256_init(&G_context.tx_info.sha256);
+        cx_hash(&G_context.tx_info.sha256.header, CX_LAST, G_context.tx_info.sighash_all, 32, G_context.tx_info.sighash_all, 32);
+        // now get second sha256
+    }
+
+    uint8_t out[256] = {0};
+    size_t sig_size = cx_ecdsa_sign(&private_key, CX_LAST | CX_RND_RFC6979, CX_SHA256, G_context.tx_info.sighash_all, 32, out, 256, NULL);
+
+    explicit_bzero(&private_key, sizeof(private_key));
+    explicit_bzero(&public_key, sizeof(public_key));
+
+    // exchange signature
+    // io_send_response < 0 means faillure
+    return io_send_response(&(const buffer_t){.ptr = out, .size = sig_size, .offset = 0}, SW_OK) >= 0;
+}
+
+void _decode_next_element() {
+    // test if there are more tokens, inputs and outputs in this order
+    // if there are, try to read from buffer
+    // if any are incomplete, inform caller to send more data 
+    if (G_context.tx_info.remaining_tokens > 0) {
+        // can read token?
+        if (G_context.tx_info.buffer_len < TOKEN_UID_LEN) {
+            // still decoding tokens, but this one was divided between calls 
+            THROW(TX_STATE_PARTIAL);
+        }
+        // for now we ignore it
```

Comment:
Ignore what?

Reply:
Add a comment explaining this condition here, please.

Reply:
Done.

Reply:
Thanks!